### DISCUSSION TITLE
JS 기본 연산자

### DISCUSSION THREAD

**Discussion by WooGie911** [wanted-9th-3team/frontend-study#58] [프론트 기술 질문 공간] [answered]
**- 아래 코드가 실행된 후, 변수 a, b, c, d엔 각각 어떤 값들이 저장될까요?**

`let a = 1, b = 1;`

`let c = ++a; `
`let d = b++; `

`a=? , b=? , c=? , d=?`

**- 아래 표현식들의 결과를 예측해 보세요.**

1. ` "" + 1 + 0`
2. `"" - 1 + 0`
3. `true + false`
4. `6 / "3"`
5. `"2" * "3"`
6. `4 + 5 + "px"`
7. `"$" + 4 + 5`
8. `"4" - 2`
9. `"4px" - 2`
10. `7 / 0`
11. `"  -9  " + 5`
12. `"  -9  " - 5`
13. `null + 1`
14. `undefined + 1`
15. `" \t \n" - 2`

**Comment by michoball** [comment]
a = 2, b = 2, c = 2, d = 1
 
 1. '10'  
 2. -1 
 3.  1
 4.  2
 5.  6
 6.  '9px'
 7.  '$45'
 8.  2
 9.  NaN
 10.  Infinity
 11.  '    -9   5'
 12.  -14
 13.  1
 14.  NaN
 15.  -2

**Reply by javascriptbiggosoo** [reply]
==를 지양해야하는 이유네요

**Comment by WooGie911** [comment, marked as answer]
- a = 2, b = 2, c = 2, d = 1

-
1. "" + 1 + 0 = "10" // (1)
2. "" - 1 + 0 = -1 // (2)
3. true + false = 1
4. 6 / "3" = 2
5. "2" * "3" = 6
6. 4 + 5 + "px" = "9px"
7. "$" + 4 + 5 = "$45"
8. "4" - 2 = 2
9. "4px" - 2 = NaN
10. 7 / 0 = Infinity
11. "  -9  " + 5 = "  -9  5" // (3)
12. "  -9  " - 5 = -14 // (4)
13. null + 1 = 1 // (5)
14. undefined + 1 = NaN // (6)
15. " \t \n" - 2 = -2 // (7)

(1). 피 연산자 중 하나가 문자열인 "" + 1에서 1은 문자형으로 변환됩니다. 따라서 공백과 문자열 1을 더한, "" + 1 = "1"과 같은 효과를 발휘하죠. 그다음 연산 "1" + 0에도 같은 규칙이 적용됩니다.
(2). 뺄셈 연산자 -는 기타 수학 연산자처럼 숫자형만을 인수로 받습니다. 빈 문자열 ""는 숫자 0으로 변환되기 때문에 결과는 -1이 됩니다.
(3). 피 연산자 중 하나가 문자열이므로 숫자 5가 문자열로 변환됩니다.
(4). 뺄셈 연산자는 인수를 숫자형으로 변화시키므로 " -9 "는 숫자 -9로 변합니다. 앞, 뒤 공백은 제거되죠.
(5). 숫자형으로 변환 시 null은 0이 됩니다.
(6). undefined는 숫자형으로 변환시 NaN이 됩니다.
(7). 문자열이 숫자형으로 변할 땐 문자열 앞뒤의 공백이 삭제됩니다. 뺄셈 연산자 앞의 피연산자는 공백을 만드는 문자 \t와 \n, 그 사이의 “일반적인” 공백으로 구성됩니다. 따라서 " \t \n"는 숫자형으로 변환 시 길이가 0인 문자열로 취급되어 숫자 0이 됩니다.

**Comment by BaekYuri** [comment]
a=2 b=2 c=2 d=1


1. 10
2. -1
3. true(1) + false(0) = 1
3. 2
4. 6
5. 9px
6. $45
7. 2
9. NaN
10. Infinity
11. -9 5
12. -14
13. null(0) + 1 = 1
14. undefined(NaN) + 1 = NaN
15. -2
